src/app/credExplorer/weights/weights.js

```diff
@@ -0,0 +1,74 @@
+// @flow
+
+import {NodeAddress, EdgeAddress} from "../../../core/graph";
+import type {NodeType, EdgeType} from "../../adapters/pluginAdapter";
+import type {StaticPluginAdapter} from "../../adapters/pluginAdapter";
+import type {StaticAdapterSet} from "../../adapters/adapterSet";
+
+export type WeightedNodeType = {|+type: NodeType, +weight: number|};
+
+export type WeightedEdgeType = {|
+  +type: EdgeType,
+  +forwardWeight: number,
+  +backwardWeight: number,
+|};
+
+export type WeightedTypes = {|
+  +nodes: $ReadOnlyArray<WeightedNodeType>,
+  +edges: $ReadOnlyArray<WeightedEdgeType>,
+|};
+
+export function defaultWeightedNodeType(type: NodeType): WeightedNodeType {
+  return {
+    type,
+    weight: type.defaultWeight,
+  };
+}
+
+export function defaultWeightedEdgeType(type: EdgeType): WeightedEdgeType {
+  return {
+    type,
+    forwardWeight: type.defaultForwardWeight,
+    backwardWeight: type.defaultBackwardWeight,
+  };
+}
+
+export function defaultWeightsForAdapter(
+  adapter: StaticPluginAdapter
+): WeightedTypes {
+  return {
+    nodes: adapter.nodeTypes().map(defaultWeightedNodeType),
+    edges: adapter.edgeTypes().map(defaultWeightedEdgeType),
+  };
+}
+
+export function combineWeights(
+  ws: $ReadOnlyArray<WeightedTypes>
+): WeightedTypes {
+  const seenPrefixes = new Set();
+  const nodes = [].concat(...ws.map((x) => x.nodes));
+  for (const {
+    type: {prefix},
+  } of nodes) {
+    if (seenPrefixes.has(prefix)) {
```

Comment:
We check for duplicate prefix, but not for duplicate name. Isn’t this
inconsistent with the adapter sets?

Reply:
No. Adapter set requires that plugin *names* be distinct, but not that their types' names be unique.

I.e. the Git Plugin and Phabricator plugin can both have a node type called 'issue'. Attempting to require that every node type name be unique across all plugins would not be feasible. Likewise for edges.

Reply:
Right. Makes sense. Thanks.